src/test/python/test_pythonparser_3.py

```diff
@@ -0,0 +1,155 @@
+import os
+import pytest
+from xml.etree import ElementTree as ET
+from src.main.python.pythonparser_3 import *
+
+_input_dir = "src/test/python/inputs/code_samples/"
+
+
+@pytest.mark.parametrize(
+    "test_input_dirs",
+    [
+        _input_dir + "small_samples",
+
+        _input_dir + "big_samples",
+
+        _input_dir + "class_samples",
+
+        _input_dir + "corner_cases_samples",
+
+        pytest.param(_input_dir + "fstring_samples",
+                     marks=pytest.mark.skip(reason='fstrings are not supported yet')),
+
+        pytest.param(_input_dir + "python_3_8_samples",
+                     marks=pytest.mark.skip(reason='python 3.8 not supported yet'))
+    ]
+)
+def test_produce_valid_xml(test_input_dirs):
+    """
+    :arg: list of input dirs (pytest parametrize)
+    Checks if parser producing valid xml file
+    :return: None
+    """
+
+    for entry in os.scandir(test_input_dirs):
+        print(f'Current entry is: {entry.path}')
+        parsed_json = parse_file(entry.path)
+        xml_str = json2xml(parsed_json)
+        try:
+            _ = ET.fromstring(xml_str)
+        except ET.ParseError as e:
+            pytest.fail(f"test file: {entry.path}, ET.ParseError msg: {e}")
+
+
+@pytest.mark.parametrize(
+    "test_input_dirs",
+    [
+        (_input_dir + "non_valid_python_code_samples")
+    ]
+)
+def test_non_valid_python_code(test_input_dirs):
+    """
+    Checks behaviour on non valid python code inputs.
+    Expected behaviour: SyntaxError is thrown during parsing
+
+    :return: None
+    """
+    for entry in os.scandir(test_input_dirs):
+        print(f'Current entry is: {entry.path}')
+        with pytest.raises(SyntaxError) as exc_info:
+            _ = parse_file(entry.path)
+        assert "invalid syntax" in str(exc_info.value)
+
+
+@pytest.mark.parametrize(
+    "input_file_paths,"
+    "expected,"
+    "xml_elements",
+    [
+        (
+            _input_dir + "small_samples/small_3.py",
+            (
+             {"lineno": "1",  "col": "0",  "end_line_no": "21", "end_col": "63"},
+             {"lineno": "13", "col": "18", "end_line_no": "13", "end_col": "27"},
+             {"lineno": "13", "col": "18", "end_line_no": "13", "end_col": "22"}
+            ),
+            (
+             "FunctionDef",
+             "BinOpLShift",
+             "SubscriptLoad"
+            )
+        )
+    ]
+)
```

Comment:
Thanks, it’s interesting suggestion, but as I understood, pytest.fixtures primarily make sense if data in fixture reused in several tests, which isn’t our case: for each unit test I have different input data. 

Moreover, I seems to me that using both fixtures and const variables for input arrays, as you suggested above, will look strange from the point of view of style. 
 
And it also seems to me that it’s better using either only fixtures or parametrization( maybe with const arrays on the top of the file)
What do you think?

Reply:
> pytest.fixtures primarily make sense if data in fixture reused in several tests 

It is not necessary. I think It is ok if you use the different approaches in the different tests if it needs. I just would like to show that `pytest.fixtures`, in this case, helps to make code better. Because at the moment the array includes the tuple which includes one string, a tuple with three dictionaries and a tuple with three strings - the more nesting and heterogeneity of the params the more difficult to read code. But with `pytest.fixtures` the array includes the tuples, all of them have the same structure and if it is need you can add names for each case in the id section - It lowered the level of nesting: 

```
    @staticmethod
    @pytest.fixture(scope="function",
                    params=[
                        ({"lineno": "1",  "col": "0",  "end_line_no": "21", "end_col": "63"}, "FunctionDef"),
                        ({"lineno": "13", "col": "18", "end_line_no": "13", "end_col": "27"},  "BinOpLShift"),
                        ({"lineno": "13", "col": "18", "end_line_no": "13", "end_col": "22"},  "SubscriptLoad")
                    ],
                    ids=[the names of the tests, if it is nessesary]
                    )
    def param_test(request) -> Tuple[Dictionary[str, str], str]:
        return request.param

    def test_correct_token_params(self, param_test) -> None:
         input_file_paths =  os.path.join(_input_dir, "small_samples/small_3.py")
        (expected, xml_elements) = param_test
        TODO
```

Is it a good approach? What do you think? I'm not completely sure that I was able to clearly show my point

Reply:
Thank you for example! Your code looks more understandable than mine. So, I will try combine fixtures and const variables to make the whole tests more clear.